### DISCUSSION TITLE
Socket status for D-Link DSP-W245

### DISCUSSION THREAD

**Discussion by pepeluga** [jonassjoh/dspW245#8] [Q&A]
Hi,
I recently bought the D-Link DSP-W245 power strip and, after searching a bit, I found this library, with which, I thought I could make a small project for home.

I'm still new to all this but looking a bit at the usage section and the code itself, I was wondering if there was any way to know the status of the sockets. Similar to how they turn on, sp.set_socket (1, on = True), but something like, status_1 = sp.get_socket (1).

I would like to know if this is possible before starting and thus know all the possibilities that this library offers.

Thanks in advance.

**Comment by Garfonso** [comment]
Hi,

In the API there is a `get_setting` command very similar to the `set_setting` command that the library sends **[here](https://github.com/jonassjoh/dspW245/blob/master/dspW245.py#L317)** .

I implemented that in my [Javascript library](https://github.com/Garfonso/dlinkWebSocketClient) (which is not tested with DSP-W245, only DSP-W115, which has only one socket). The message send for `get_setting` is much easier, like [here](https://github.com/Garfonso/dlinkWebSocketClient/blob/master/index.js#L276) (where type is 16 for socket and 41 for LED) and you'd probably need to set the idx for dsp-w245, too. (as said, not tested in my library, sadly. If you want to test / extend, I'm willing to either take a PR or try to do it with you testing stuff).

But experimenting with that library, I noticed, that the device also sends a push message if the socket changes (which is handled [here](https://github.com/Garfonso/dlinkWebSocketClient/blob/master/index.js#L73), if you know a bit Javascript).

**Reply by pepeluga** [reply]
Hi again,

I have tried to get the state with the function you said but it returns None when displayed on the screen.

The rest of the example of the usage section works but the way I want to get the status, as I said, it returns None. I don´t know if I using it the wrong way and sorry if this is a stupid question, I just wanted to try to do something with this.


from dspW245 import SmartPlug

sp = SmartPlug("iponmynet", "pinofmydspW245")

sp.set_socket(1, on=True)
sp.set_socket(4, on=True)

status = sp.send_get_setup_status()
print(status)         <---- None 

sp.close()

By the way, thanks for responding so fast and I also took a look at your Javascript library but as I am still learning I feel more comfortable with Python and creating simple scripts to do things like this that I am trying.

**Reply by jonassjoh** [reply]
I don't even remember adding the `send_get_setup_status` function, but unless you've modified it yourself, the function does not return anything and will thus always print "None".
Try modifying the function to return line `337` or try `sp = SmartPlug(HOST, pin, verbose=2)` to get more verbose output.

**Reply by lordmane** [reply]
You can use a new function to retrieve the status of the leds (0=all leds status):

    def get_led(self, led):
        """
        On the DSP-W245 led can be 1,2,3,4.
        """
        data = {
                "command": "get_setting",
                "setting":[
                {
                        "uid": 0,
                        "metadata": {
                           "value": 1
                    },
                        "name ":"DSP-{}-{}-{}".format(self.model, self.u_nr(), str(led)),
                        "idx": (led - 1),
                        "type": 16
                }
            ]
        }
        resp = self.send_json(data, log="Getting LED #{}".format(led))
        return resp
